app/scripts/controllers/network/provider-api-tests/shared-tests.js

```diff
@@ -674,61 +787,70 @@ export function testsForRpcMethodSupportingBlockParam(
           response: { result: mockResults[0] },
         });
 
-        const results = await withInfuraClient(({ makeRpcCallsInSeries }) =>
-          makeRpcCallsInSeries(requests),
+        const results = await withClient(
+          { type: providerType },
+          ({ makeRpcCallsInSeries }) => makeRpcCallsInSeries(requests),
         );
 
         expect(results).toStrictEqual([mockResults[0], mockResults[0]]);
       });
     });
 
-    it('hits Infura and does not reuse the result of a previous request if the latest block number was updated since', async () => {
-      const requests = [
-        { method, params },
-        { method, params },
-      ];
-      const mockResults = ['first result', 'second result'];
+    if (providerType === 'infura') {
```

Comment:
It looks like this is exposing a genuine bug in our custom RPC provider, specifically in the cases where the block tag is unset or set to `latest`.

Reply:
cc @mcmire, it looks like if I replace `BlockRefRewriteMiddleware` with `BlockRefMiddleware` in the network client, this bug is fixed. I'm guessing what's happening here is that `BlockRefRewriteMiddleware` is replacing `latest` with the _old_ block number without first checking the block tracker. Then the next middleware, the block cache middleware, is returning the stale cache result.

Reply:
Hmm, interesting. Okay I'll take a closer look at this when I get a chance.

Reply:
I can create a PR to fix this bug in the custom RPC client. I just tagged you because of how this relates to work you've done in the `eth-json-rpc-middleware` package. I remember you had at one point suggested that we get rid of the "child request" in `BlockRefMiddleware`, but I suspect doing that would have introduced this bug in that middleware as well.

If that is the case, we should add a test to `eth-json-rpc-middleware` to capture this behavior and delete the broken `BlockRefRewriteMiddleware` from the package completely.

Reply:
Nevermind. This was a mistake in the tests. The `params` are being shared between the two calls here, so when that object is directly mutated in one it gets changed in the second call as well.

The middleware itself is working just fine. They both correctly ask the block tracker for the latest block number.